File: src/com/kingmang/lazurite/parser/ast/WhileStatement.java
package com.kingmang.lazurite.parser.ast;


public final class WhileStatement extends InterruptableNode implements Statement {
    
    public final Expression condition;
    public final Statement statement;

    public WhileStatement(Expression condition, Statement statement) {
        this.condition = condition;
        this.statement = statement;
    }
    
    @Override
    public void execute() {
        super.interruptionCheck();
        while (condition.eval().asInt() != 0) {
            try {
                statement.execute();
            } catch (BreakStatement bs) {
                break;
            } catch (ContinueStatement cs) {
                // continue;
            }
        }
    }
    
    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }

    @Override
    public <R, T> R accept(ResultVisitor<R, T> visitor, T t) {
        return visitor.visit(this, t);
    }

    @Override
    public String toString() {
        return "while " + condition + " " + statement;
    }
}

File: src/main/java/com/kingmang/lazurite/parser/ast/statements/WhileStatement.kt
package com.kingmang.lazurite.parser.ast.statements

import com.kingmang.lazurite.parser.ast.InterruptableNode
import com.kingmang.lazurite.parser.ast.expressions.Expression
import com.kingmang.lazurite.patterns.visitor.ResultVisitor
import com.kingmang.lazurite.patterns.visitor.Visitor

data class WhileStatement(
    val condition: Expression,
    val statement: Statement
) : InterruptableNode(), Statement {
    override fun execute() {
        super.interruptionCheck()
        while (this.condition.eval().asInt() != 0) {
            try {
                this.statement.execute()
            } catch (bs: BreakStatement) {
                break
            } catch (cs: ContinueStatement) {
                // continue;
            }
        }
    }

    override fun accept(visitor: Visitor) =
        visitor.visit(this)

    override fun <R, T> accept(visitor: ResultVisitor<R, T>, input: T): R? =
        visitor.visit(this, input)

    override fun toString(): String =
        "while ${this.condition} ${this.statement}"
}
